### ISSUE TITLE
SL5/SP9 does not power off properly

### ISSUE THREAD

**Issue by qzed** [linux-surface/linux-surface#1058]
The Intel Surface Pro 9 and Surface Laptop 5 do not power off properly. Reboot seems to work fine. See #1010 and [wiki](https://github.com/linux-surface/linux-surface/wiki/Intel-12th-Generation-Devices-Issues#s5-acpi-shutdown-or-regular-shutdown-does-not-work).

**Comment by qzed** [comment]
I've received some reports via mail and here's the summary: Testing via [this module](https://github.com/linux-surface/surface-shutdown) (and variations) indicates that the issue is some problem in the firmware. In particular: When inserting different shutdown procedures as highest priority, the system

- locks up with
  ```c
  efi.reset_system(EFI_RESET_SHUTDOWN, EFI_SUCCESS, 0, NULL);
  ```
- reboots with
  ```c
  efi.reset_system(EFI_RESET_COLD, EFI_SUCCESS, 0, NULL);
  ```
  or
  ```c
  efi.reset_system(EFI_RESET_WARM, EFI_SUCCESS, 0, NULL);
  ```
  and also with
  ```c
  efi.reset_system(EFI_RESET_SHUTDOWN, EFI_SUCCESS, 0, NULL);
  efi.reset_system(EFI_RESET_WARM, EFI_SUCCESS, 0, NULL);
  ```

That indicates that `EFI_RESET_SHUTDOWN` returns (or fails internally in some way that makes it return). The lock-up is then probably caused by some fallback, potentially even some endless loop. From what I can tell, by default there should only be the ACPI power-off handler installed on this system. My guess is that this might call `reset_system()` internally.

Some ideas:
- Courtesy of @StollD: We could maybe use EFI variables to store some debug info...
- Call `reset_system(EFI_RESET_SHUTDOWN)` directly at runtime and see what happens. No idea if that would work properly under normal circumstances without any sort of preparation...
- On the SPX, `reset_system()` fails due to a page fault (regardless of the mode). So it might be worth looking at the [page fault handling](https://elixir.bootlin.com/linux/v6.1.12/source/arch/x86/platform/efi/quirks.c#L747). in particular, removing the `machine_real_restart(MRR_BIOS)` call and then just calling `reset_system()` at runtime might at least let us know if a page fault occured.

**Comment by qzed** [comment]
Something more concrete to test: Since the module returns, can someone try replacing the [content of `surface_shutdown_init()`](surface_shutdown_init) with
```c
pr_warn("Calling efi.reset_system(EFI_RESET_SHUTDOWN, ...)");
efi.reset_system(EFI_RESET_SHUTDOWN, EFI_SUCCESS, 0, NULL);
pr_warn("Call to efi.reset_system(EFI_RESET_SHUTDOWN, ...) returned");
return 0; 
```
And see if a) the system keeps running and b) whether there is some error message or warning in the kernel log. Ideally, boot with `efi=debug` for that.

**Comment by WolfWings** [comment]
Sorry to bumble into another issue with my SG3: I've been seeing similar issues where the ACPI shutdown results in a simply hung tablet after it's been running for several hours so it's slept/woken up a few times, and would keep running the battery down silently unless I held the power button for 30+ seconds to hard poweroff the device.

ACPI reboot would also take several seconds of a black screen in these cases, however when I tried `reboot=efi` I've never had it take more than maybe a second to finish the reboot, and adding a minimal 'poweroff.efi' compiled using gnu-efi to my bootloader so I can simply 'reboot into EFI poweroff' for the time being was my workaround while I was debugging further.

So there's likely something that's being handled incorrectly at some point in the ACPI and/or EFI system calls during normal operation that's bungling up the direct shutdown process on these newer systems and leaving things in a weird state?

For reboot the fallback tree is documented at https://git.kernel.org/pub/scm/linux/kernel/git/stable/linux.git/tree/arch/x86/kernel/reboot.c?h=v6.1.12#n566

But for poweroff the logic is (while spread out like hayseed) very simple when ACPI is available:

1. If S5 is unavailable OR it's "ACPI Reduced Hardware" then it attempts EFI
2. ACPI
3. pm_power_off
4. The final fallback is the 'tboot' function chains with the end of this is a literal `while (1) halt();` eventually

EFI logic starts in [/drivers/firmware/efi/reboot.c](https://git.kernel.org/pub/scm/linux/kernel/git/stable/linux.git/tree/drivers/firmware/efi/reboot.c?h=v6.1.12#n49)

That gets overridden in [/arch/x86/platform/efi/quirks.c](https://git.kernel.org/pub/scm/linux/kernel/git/stable/linux.git/tree/arch/x86/platform/efi/quirks.c?h=v6.1.12#n596) to add the S5/ACPI Reduced Hardware check.

The reduced hardware check requires the ACPI FADT table which at least my Surface Go 3 doesn't have, in [/drivers/acpi/acpica/tbfadt.c](https://git.kernel.org/pub/scm/linux/kernel/git/stable/linux.git/tree/drivers/acpi/acpica/tbfadt.c?h=v6.1.12#n381)

The S5 check is in [/drivers/acpi/sleep.c](https://git.kernel.org/pub/scm/linux/kernel/git/stable/linux.git/tree/drivers/acpi/sleep.c?h=v6.1.12#n1085)

pm_power_off is being removed from the kernel but it LITTERS the thing for legacy hardware; I don't see any relevant for my SG3 at least? But there's thousands of entries for that so I might've missed one.

The actual shutdown chain based on priority gets called from [/arch/x86/kernel/reboot.c](https://git.kernel.org/pub/scm/linux/kernel/git/stable/linux.git/tree/arch/x86/kernel/reboot.c?h=v6.1.12#n740), with tboot.c code chains as the fallback.

I'll poke at getting [ramoops](https://git.kernel.org/pub/scm/linux/kernel/git/stable/linux.git/tree/Documentation/admin-guide/ramoops.rst?h=v6.1.12) working this weekend, unless PStore has been found to have issues on any Surface device in the past?

**Comment by qzed** [comment]
@WolfWings This is very likely a different root problem. Can you open a new issue? I'll respond and answer to your questions there.

On the SP9, we've now nailed this down to some PCI-related code that gets run on shutdown and breaks EFI's ResetSystem call. I'll write a more detailed update here shortly.

**Comment by qzed** [comment]
We managed to track down this issue to a fairly limited scope yesterday on our Matrix chat.

It turns out that EFI's ResetSystem works fine before calling some shutdown related code. In particular, calling it right at the beginning of [`kernel_power_off()`](https://elixir.bootlin.com/linux/v6.1.12/source/kernel/reboot.c#L678) works fine. This means that ResetSystem is generally working fine unless some conditions are met.

We have further tracked the issue down to some PCI related shutdown code. In particular, applying the following diff makes a normal shutdown (`shutdown --poweroff now`) work:

```diff
diff --git a/drivers/base/core.c b/drivers/base/core.c
index d02501933467..032ed81a09fc 100644
--- a/drivers/base/core.c
+++ b/drivers/base/core.c
@@ -30,6 +30,7 @@
 #include <linux/swiotlb.h>
 #include <linux/sysfs.h>
 #include <linux/dma-map-ops.h> /* for dma_default_coherent */
+#include <linux/pci.h>
 
 #include "base.h"
 #include "physical_location.h"
@@ -4658,7 +4659,8 @@ void device_shutdown(void)
 		if (dev->bus && dev->bus->shutdown) {
 			if (initcall_debug)
 				dev_info(dev, "shutdown\n");
-			dev->bus->shutdown(dev);
+			if (dev->bus != &pci_bus_type)
+				dev->bus->shutdown(dev);
 		} else if (dev->driver && dev->driver->shutdown) {
 			if (initcall_debug)
 				dev_info(dev, "shutdown\n");
```

Note that this should be a shutdown via ACPI (we don't explicitly call ResetSystem), so it's likely that ACPI code calls ResetSystem internally, or does something to a similar effect that breaks in the same (or a similar) way.

We were able to further limit the scope of this via the following diff:

```diff
diff --git a/drivers/pci/pci-driver.c b/drivers/pci/pci-driver.c
index 107d77f3c846..0892f12d3ba7 100644
--- a/drivers/pci/pci-driver.c
+++ b/drivers/pci/pci-driver.c
@@ -509,8 +509,8 @@ static void pci_device_shutdown(struct device *dev)
 
 	pm_runtime_resume(dev);
 
-	if (drv && drv->shutdown)
-		drv->shutdown(pci_dev);
+//	if (drv && drv->shutdown)
+//		drv->shutdown(pci_dev);
 
 	/*
 	 * If this is a kexec reboot, turn off Bus Master bit on the
```

Applying this diff alone does _not_ make it work.

Therefore, the the problem must lie in [`pci_device_shutdown()`](https://elixir.bootlin.com/linux/v6.1.12/source/drivers/pci/pci-driver.c#L505) (diff 1) but outside of the driver-specific `drv->shutdown()` callback (diff 2).

**Comment by qzed** [comment]
Limiting it further: The issue is likely caused by one (or multiple) of the following devices (SP9):
```
i915            0000:00:02.0
pcieport        0000:00:06.0
thunderbolt     0000:00:0d.3
xhci_hcd        0000:00:14.0
mei_me          0000:00:16.0
nvme            0000:02:00.0
pcieport        0000:00:07.0
pcieport        0000:00:07.3
xhci_hcd        0000:00:0d.0
snd_hda_intel   0000:00:1f.3
```

**Comment by qzed** [comment]
Another write-up from our findings in Matrix chat:

First off, we have a workaround: https://github.com/linux-surface/kernel/commit/6f57d3f0780712e9b6cea1ee3b87605d2920e8ae.

The problem seems to be a combination of the `pm_runtime_resume()` and the `drv->shutdown()` call, as indicated in my previous comment. Just commenting out/removing one of those won't work. Furthermore, this is not tied to a single PCI device, but instead it seems that we must not runtime-resume and `drv->shutdown()` multiple devices. The smallest list of devices to exclude from that that we have found on the Surface Pro 9 is

| Driver     | VID/PID     | Domain/Bus/Slot/Fn | Comment                             |
|------------|-------------|--------------------|-------------------------------------|
| `i915`     | `8086:46a8` | `0000:00:02.0`     | iGPU                                |
| `pcieport` | `8086:462f` | `0000:00:07.0`     | Thunderbolt 4 PCI Express Root Port |
| `pcieport` | `8086:462f` | `0000:00:07.3`     | Thunderbolt 4 PCI Express Root Port |
| `xhci_hcd` | `8086:461e` | `0000:00:0d.0`     | Thunderbolt 4 USB Controller        |

It may be possible that this list is not exhaustive. If you still have problems with shutdown, please try whether adding a device to the workaround linked above (i.e. adding the respective `DECLARE_PCI_FIXUP_FINAL` statement for `quirk_no_shutdown`) fixes this.

This list may also not match the SL5, due to which that is currently excluded from the workaround linked above. I would appreciate if someone can test this before I enable the quirk on the SL5. Feedback is appreciated.

**Comment by qzed** [comment]
Forgot to mention: Thanks go to @aj3423 and `nouknouk` from our Matrix chat for debugging and testing all of this.

**Comment by CheeseFM** [comment]
Heya! I own a Surface Laptop 5 with this issue and am willing to help, how do I go about this?

**Comment by qzed** [comment]
Do you have any experience building custom kernels? Also, feel free to hop on our [Matrix chat](https://matrix.to/#/#linux-surface-support:matrix.org) for this.

**Comment by CheeseFM** [comment]
I have some experience but I wouldn't call it a lot, I'm very willing to learn! Downloading a matrix client right now...

**Comment by qzed** [comment]
Alright, let me know when you're ready.

**Comment by CheeseFM** [comment]
Installing fedora right now, I'm guessing I'll also need the linux-surface kernel installed to start?

**Comment by qzed** [comment]
Not necessarily for testing, but you probably want to do that anyways. At least if you want touch support. Type-cover support should already work on the 6.1 kernels that fedora provides.

**Comment by qzed** [comment]
Thanks to @CheeseFM shutdown should work with the next release on the SL5 as well. We have to add the Thunderbolt 4 NHI to the list as well. I've updated my comment above.

**Comment by vortarian** [comment]
FWIW, I'm using a Surface Laptop 5 model 1979 (15/i7/32gb), and the camera, shutdown and suspend seem to be working perfectly.  Using Kubuntu build:
`Linux codez 6.2.5-surface #1 SMP PREEMPT_DYNAMIC Sun Mar 12 17:50:04 UTC 2023 x86_64 x86_64 x86_64 GNU/Linux`
I followed the default instructions on linux-surface/ubuntu install, and have turned on the Microsoft/3rd Party CA for Secure Boot.

**Comment by LubosD** [comment]
@WolfWings Have you created that issue for SG3? I've noticed my SG3 doesn't really turn off with the latest surface kernel when I do a shutdown. It just heats up the device until the battery dies.

**Comment by WolfWings** [comment]
@LubosD No, I'm sorry work had me travelling extensively the last few weeks so I'd never circled back to the poweroff problem for the SGO3 to create a separate issue yet.

**Comment by PhilDevProg** [comment]
@LubosD @WolfWings I experienced this too and created an issue: https://github.com/linux-surface/linux-surface/issues/1111

**Comment by gafenton** [comment]
I actually had no problems shutting down my SP9 until I upgraded to kernel 6.2.10-200.fc37.x86_64 a few days ago. Now it doesn't properly shutdown and I have to hold down the power button for 20+ seconds to wake it up, boot into Windows, and shutdown properly.
Same problem with 6.2.12-200.fc37 (just installed now).
I'm pretty sure that 6.2.8-200.fc37 was working, if that helps.

**Comment by idrisr** [comment]
Hi all, I'm running a Surface Pro 9 and would be willing to help debug this with the custom kernel as long I could get some help on the steps. Currently I'm running nixos but I can install something else for the debugging -- then porting the fix to nixos should not be too hard. Happy to help and thanks for all the efforts to date!

**Comment by 0none** [comment]
Just want to add some anecdotal data for my SL5:
Shutting down seems to work with the 6.4.12-arch1-1-surface kernel.
Suspending is inconsistent, and only seems to work every other time?

**Comment by gafenton** [comment]
Hi,
      
         I just wanted to say thanks very much for this pointer and fix!
      My SP9 is now shutting down correctly using
      6.4.7-1.surface.fc38.x86_64. Not quite the same as yours, but
      certainly does the trick. I haven't tried suspending. I don't
      usually bother suspending. I want my laptop completely off, so
      that when I come back to it days, or weeks, later, the battery is
      still fully charged.
      
         Best wishes,
           Gordon
      
      PS: I'm sorry if I'm using the wrong channel to send my thanks.
      
      PPS: now I just need to figure out if there is an easy way of
      preventing "dnf -y upgrade" from replacing that kernel.
    
    On 9/15/23 06:34, theo wrote:
    
    
      
      
      
        Just want to add some anecdotal data for my SL5:
          Shutting down seems to work with the 6.4.12-arch1-1-surface
          kernel.
          Suspending is inconsistent, and only seems to work every other
          time?
        —
          Reply to this email directly, 
            view it on GitHub, or 
            unsubscribe.
          You are receiving this because you commented.Message
            ID: ***@***.***>
        [
{
***@***.***": "http://schema.org",
***@***.***": "EmailMessage",
"potentialAction": {
***@***.***": "ViewAction",
"target": "https://github.com/linux-surface/linux-surface/issues/1058#issuecomment-1720974493",
"url": "https://github.com/linux-surface/linux-surface/issues/1058#issuecomment-1720974493",
"name": "View Issue"
},
"description": "View this Issue on GitHub",
"publisher": {
***@***.***": "Organization",
"name": "GitHub",
"url": "https://github.com"
}
}
]

**Comment by mkober** [comment]
On the SL5 with the 6.6.11-surface-1 kernel on Ubuntu 23.10 the shutdown is working correctly, still testing the suspending.

**Comment by Svampebob1** [comment]
> On the SL5 with the 6.6.11-surface-1 kernel on Ubuntu 23.10 the shutdown is working correctly, still testing the suspending.

I'm receiving a SP5 tomorrow. It seems SL5 is very similar? How is the suspend working for you?

**Comment by derkrasseleo** [comment]
My SP9 also doesn't shutdown/reboot properly without holding the power button for about 10 seconds..

**Comment by gafenton** [comment]
My SP9 was having the same trouble when I tried to shutdown from Fedora. But installing the 6.4.7-1.surface.fc38.x86_64 kernel fixed that problem. (There is a more recent 6.7.2-1.surface kernel available but I haven't had a chance to try it out... see why next.)

More recently, I made the mistake of allowing Win11 to update the firmware and now I can't boot into Fedora at all, even from a live USB stick. There is a nice summary of the problem on

https://github.com/linux-surface/linux-surface/issues/1162

Looks like I'll have to wait for Fedora 40 (or so) before I can use Fedora on this SP9. Probably wise not to let Windows update the firmware!

**Comment by derkrasseleo** [comment]
Strange, I also remember a firmware update on windows a while ago. I am on the newest kernel btw. (linux-surface 6.7.6)

**Comment by derkrasseleo** [comment]
I can confirm I can shut down and restart from GNOME 46 and with kernel 6.8.1. It still hangs for a minute or so, but it eventually shuts down. Also, no more kernel panic or weird logs, only blinking cursor.

**Comment by zoukyle** [comment]
My SP9 works for shutdown, but hibernation doesn't work. After hibernation, the computer won't respond to the power button any more and I have to hold for 10+ seconds to reset to reboot.

**Comment by derkrasseleo** [comment]
> My SP9 works for shutdown, but hibernation doesn't work. After hibernation, the computer won't respond to the power button any more and I have to hold for 10+ seconds to reset to reboot.

Can confirm, but it doesn't seem consistent, sometimes (when I don't close the lid/case) it wakes up normally, other times it "shuts down"

**Comment by bastianandre-myenso** [comment]
Can confirm sporadic issues with shutdown on ubuntu 22.04 with linux surface kernel 6.8.8 on SP9.
Have not tried the workaround mentioned by qzed though.

**Comment by 3xfc** [comment]
Shutdown seems to be working on my SP9 when running a mainline kernel with the option `pci=hpiosize=0`. See #1082.

**Comment by stlaz** [comment]
This seems to be a problem on SL6, too.

An endless loop as suggested in first comments might be the case here - the laptop's battery got drained overnight, I didn't notice the laptop was still on.

**Comment by nouknouk** [comment]
Hello,

I observe now a strange behavior that wasn't observed months ago:

My SP9 is setup with a (grub) dual boot:
- win 11 pro 23H2
- linux mint with surface kernel 6.10.5

When I boit under linux, no issue: I can shutdown or reboot without issue.

When I boot under windows, an attempt to shutdown from windows doesn't shutdown: the screen comes black for a few seconds, and then the login screen shows again (without a reboot, as I didn't saw the grub page in-between). Also the windows hello auth via webcam doesn't work anymore.
On the opposite, a restart action from windows behaves well and I get the grub menu.

That means after a windows session, in order to shutdown my SP9, I need first to make a reboot from windows to restart a linux session, and then trigger a shutdown from Linux.

**Comment by sh0x31** [comment]
Hi,
I upgraded from the Surface Laptop Go to the Surface Laptop 5. Are there any updates on the issues with Intel 12th Gen CPUs? Can I follow the Surface Pro 9 guides for machine power state/operation and ACPI interrupt storm? I read somewhere here that 6.12 solves some of these issue, is that true? Maybe I’ll wait for the Surface Linux Kernel in 6.12 then. ( :
Thanks for sharing your experience!

**Comment by petm5** [comment]
@sh0x31 Shutdown should work fine with linux-surface v6.1.12-2 and above. The Surface Pro 9 guide for fixing the ACPI interrupt storm should apply to other Intel 12th generation devices too.

If you don't mind me asking, where did you hear about fixes in v6.12 for the Surface Laptop 5 or for Intel Surface devices? I believe the upcoming fixes are mostly related to Qualcomm devices like the Surface Laptop 7.

**Comment by toastedsalad** [comment]
> This seems to be a problem on SL6, too.
> 
> An endless loop as suggested in first comments might be the case here - the laptop's battery got drained overnight, I didn't notice the laptop was still on.

Can confirm happens on my SL6 as well. 
OS: Ubuntu 22
Kernel: 6.12.3-surface-2

I think I haven't noticed it initially because I would always either suspend or reboot. But yeah, it wouldn't turn on without doing the power button hold maneuver. And here too I found the battery completely drained.

**Comment by VaryTea** [comment]
(SP10) Same issue, powering off does not power off device; has lead to battery dying often. Need to hold power button until full power off.

Surface Pro 10
Kernel version 6.12.7
Arch

**Comment by goxia** [comment]
The issue also happened with my LP6
Linux 6.12.3-surface-2

**Comment by GreySimphonic** [comment]
I have a Surface Laptop 5 and am happy to help test any kind of possible validation for a fix (mentioned on the wiki page for this device). Competent enough to adjust linux-surface to compile as a UKI, not competent enough to contribute directly to the project. I think I'd be capable enough to validate a potential fix without much friction or having to overly explain/teach me what was needed. Arch user, for whatever that's worth.

Edit: Just so I don't get too excited if not appropriate, is this just (_potentially_) fixing powering all the way off, or will this also hopefully fix the inconsistent suspending when closing the lid? I'll be happy to test either way...just curious how invested I should be in being able to use this machine. With no reliable sleep on lid close the laptop is likely dead to me. Still would be happy to help validate proper poweroff for others though.

**Comment by soehner** [comment]
## Full coexistence on SP9: standby + real hibernate (with auto-poweroff) + clean shutdown

On a Surface Pro 9 (Intel), Ubuntu 26.04, kernel `6.19.8-surface-3`, LUKS root, I got all three power paths working at the same time. The individual pieces are mostly known (credit below), but the **interaction** between them isn't documented anywhere, so I want to write it up in one place.

### The key insight (this is the part I didn't find documented)

The `pcie_port_pm=off pcie_aspm=off` parameters that are needed so the machine *fully powers off after hibernate* (instead of hanging at "System halted", see #1864) are **themselves the trigger** for the GPE storm (`ACPI Error: No handler or method for GPE …`) that hangs a *normal* shutdown (S5). I confirmed this by isolation testing — each pcie param **on its own** reproduces the S5 hang:

| Poweroff attempt | cmdline params | result |
|---|---|---|
| none | — | clean ✅ |
| `acpi_osi` + both pcie | GPE storm ❌ |
| both pcie only | hang ❌ |
| `pcie_port_pm=off` only | hang ❌ |
| `pcie_aspm=off` only | hang ❌ |

The fix is the workaround @petm5 already found for the GPE spam in #1082: **`pci=hpiosize=0`**. It cures exactly the storm the pcie params provoke. Quantitative evidence on the running system:

```
# before (pcie params, no hpiosize):
grep -c invalid /sys/firmware/acpi/interrupts/gpe_all   → 119   (of 128)
# after adding pci=hpiosize=0:
grep -c invalid /sys/firmware/acpi/interrupts/gpe_all   → 0
```

So `pci=hpiosize=0` isn't just an "ACPI error spam" cosmetic fix — on the SP9 it's what lets the poweroff-enabling pcie params coexist with a clean S5 shutdown.

### Full working recipe

**1. Kernel cmdline** (`/etc/default/grub.d/…`):
```
acpi_osi="Windows 2022" pcie_port_pm=off pcie_aspm=off pci=hpiosize=0
```

**2. Hibernate mode** — the SP9 is a Modern-Standby (s2idle-only) device with no working platform-S4 path, so the platform mode never writes an image. Force the shutdown path (`/etc/systemd/sleep.conf.d/10-hibernate-shutdown.conf`):
```
[Sleep]
HibernateMode=shutdown
```

**3. resume** — `resume=UUID=…`, `resume_offset=…` for a 16 GB `/swap.img` (filefrag-verified offset).

**4. Enable hibernate for the desktop user** — Ubuntu blocks it by default in `com.ubuntu.desktop.rules`; override with `/etc/polkit-1/rules.d/00-enable-hibernate.rules` (sorts first, YES wins), then restart powerdevil.

**5. suspend-then-hibernate on battery** — KDE powerdevil `[Battery][SuspendAndShutdown] SleepMode=3`, `HibernateDelaySec=30min`. Lid-close on battery → s2idle → real hibernate after ≤30 min.

### Verified
- Standby (lid), manual `systemctl hibernate`, GUI "Hibernate" menu item, automatic battery lid→hibernate after ~30 min, **and** clean normal shutdown — all confirmed on real reboots.
- Hibernate resume proven real (same boot-ID, `hibernation entry`→`exit` with a powered-off wall-clock gap, running session restored with full context).

### Credit
`pci=hpiosize=0` is @petm5's workaround from #1082 (already in the SP9 wiki). `HibernateMode=shutdown` and the polkit override are generic (Arch wiki). What I'm adding here is the **causal link** (pcie params cause the S5 hang, hpiosize cures it), the quantitative GPE measurement, and the fact that all three power modes can coexist.

One cosmetic remainder: after a hibernate-poweroff the power button needs a ~20 s hold to wake (EC deep-sleep, firmware level, functionally irrelevant).

**Comment by User25665** [comment]
Can confirm the kernel cmdline addition fixes the shutdown on a SL5!

Unfortunately sleep is still unpredictable and Hibernate still doesn't fully shutdown. Regarding that last one, @soehner: are you sure your 20s hold is purely cosmetic? Because that's one of the symptoms how I now it doesn't reach S5 , combined with the usb ports still being powered. (Same as the original issue.)

------------------------------------

Update a day later: I've managed to fix my problems above, see https://github.com/linux-surface/linux-surface/issues/1010#issuecomment-4762977976
TL;DR: Root cause is the SOF audio DSP (PCI 8086:51c8) staying in D0 after the first resume, hanging the next suspend. Fix: add snd_intel_dspcfg.dsp_driver=1 to the kernel cmdline. Combined with the fix from soehner resolves all issues for me with Machine Power State/Operations.
